### PR TITLE
Recipes App Project: Dev to Main PR

### PR DESCRIPTION
**Pull request by Profsain**
### Recipes App Group Project: 3 members in this team
Hi reviewer, in this group project the team implemented the following requirement:
- Created a login page.
- Created a registration page.
- Created food list (with all CRUD implementation, except for 'update').
- Created recipes list (with all CRUD implementation, except for 'update').
- Implemented general shopping list view (all missing food for all your recipes and total price).
- Implemented a list of all public recipes from other users with their names and total prices.
- Created Inventories list (with all CRUD implementation, except for 'update').
- Implemented Inventory shopping list, a shopping list, but only taking into consideration a chosen recipe and inventory.
- Wrote spec test for model, request, and system
- Solved N+ error
- Fixed all linter errors
- Merged all changes to dev branch for review PR

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,158 @@
...
+  - [🚀 Live Demo](#live-demo)
+- [💻 Getting Started](#getting-started)
+  - [Setup](#setup)
+  - [Prerequisites](#prerequisites)
+  - [Install](#install)
+  - [Usage](#usage)
+  - [Run tests](#run-tests)
+  - [Deployment](#triangular_flag_on_post-deployment)
+- [👥 Authors](#authors)
+- [🔭 Future Features](#future-features)
```
Comment (Kamba56) [inline README.md:20]: - [ ] Kindly add some future features under the future features section in your readme file.

File: app/views/recipes/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    <span><%= @recipe.description %></span>
+  </div>
+  
+  <%# modal and its button %>
+  <% if can? :destroy, @recipe %>
+    <div class="flex justify-between my-2">
+      <!-- Modal toggle -->
+      <button data-modal-target="defaultModal" data-modal-toggle="defaultModal" class="block text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800" type="button">
+        Generate Shopping list
+      </button>
```
Comment (Kamba56) [inline app/views/recipes/show.html.erb:30]: - [ ] The generate shopping list button should open a modal to choose an inventory to create a shopping list with for that recipe as shown in the wireframe below. Currently, your button doesn't function at all, kindly fix that, please.
Screenshot from 2023-02-17 10-40-43

- [ ] Also add a shopping list page to match the given wireframe below. This will display the shopping list generated with the recipe name, and inventory name, Shows the list of food that is missing by comparing the food in the recipe with the food in the inventory and Counts the total food items and total price of the missing food.
image

File: app/views/public_recipes/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<h1 class="text-2xl font-extrabold text-gray-900  md:text-3xl lg:text-4xl"> Public Recipes List </h1>
+<div class="flex justify-center">
+  <%= image_tag ("recipes4.jpg"), :class => "w-9/12 border rounded-3xl"%>
+</div>
+<div class="wrapper">
+
+<% if @public_recipes.length != 0 %>
+<% @public_recipes.each do |recipe| %>
+  <%= render partial: 'public_recipe_card', locals: { recipe: recipe, user_id: recipe.id } %>
```
Comment (Kamba56) [inline app/views/public_recipes/index.html.erb:9]: - [ ] As demanded by project requirements, clicking on a public recipe should lead to the details of the recipe. Kindly fix this, please.
image
image

**Review by Kamba56 — CHANGES_REQUESTED** [review]
Hi @Profsain, @zakiyousufi, @OmarMWarraich , 

Good job so far!!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- [x] Great use of Github workflow!👏👏👏
- [x] All linters passed

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by zakiyousufi** [comment]
Hello @OmarMWarraich @Profsain 
I fixed problems please don't push anything to the branch
Just the Generate shopping list button needs a reload to work I don't know why

File: app/views/public_recipes/index.html.erb (hunk 3 above)
Comment (zakiyousufi) [reply app/views/public_recipes/index.html.erb:9]: Hello @Kamba56 
Thanks for your kind review and feedback (Click on recipe name, It will lead you to the recipe detail page)

**Comment by OmarMWarraich** [comment]
> Hello @OmarMWarraich @Profsain I fixed problems please don't push anything to the branch Just the **Generate shopping list button needs a reload to work I don't know why**

Just remove my comments and the highlighted sentence and better remove our conversation and then append the scrrenshots of the modal working and the link to recipe details in the public recipe section working. I think we would be good to go. So you would submit for review. Let us know here if you need any help. 

Maybe the button needs overloading

File: app/views/shopping_list/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      </thead>
+      <tbody>
+        <% @missing_foods.each do |food| %>
+          <tr class="bg-white border-b dark:bg-gray-900 dark:border-gray-700">
+            <th scope="row" class="px-6 py-4 font-medium text-gray-900 whitespace-nowrap dark:text-white"><%= food.name %></th>
+            <td class="px-6 py-4 font-medium text-gray-900 whitespace-nowrap dark:text-white"><%= food.quantity(@recipe) %></td>
+            <td class="px-6 py-4 font-medium text-gray-900 whitespace-nowrap dark:text-white">$ <%= food.price.to_i * food.quantity(@recipe).to_i %></td>
+          </tr>
+        <% end %>
+      </tbody>
```
Comment (VICTOR-CHUKWU) [inline app/views/shopping_list/index.html.erb]: - [ ] for some reasons this table is not showing even when I have added food to the inventory food list, kindly fix this
how shopping list page looks like :point_down: 
Screenshot from 2023-02-17 15-28-52

wireframe :point_down: 
Screenshot from 2023-02-17 15-16-35

**Comment by zakiyousufi** [comment]
> Hello @OmarMWarraich @Profsain I fixed problems please don't push anything to the branch Just the Generate shopping list button needs a reload to work I don't know why

File: app/views/recipes/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  </div>
+  <div>
+    <span class="text-lg font-semibold text-gray-900">Steps :</span>
+    <span><%= @recipe.description %></span>
+  </div>
+  <%# modal and its button %>
+  <% if can? :destroy, @recipe %>
+    <div class="flex justify-between my-2">
+      <!-- Modal toggle -->
+      <button data-modal-target="defaultModal" data-modal-toggle="defaultModal" class="block text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800" type="button">
```
Comment (VICTOR-CHUKWU) [inline app/views/recipes/show.html.erb:28]: - [ ] I feel why the modal does not pop up on initial render should be a bootstrap issues, kindly check your bootstrap version and make sure you used the correct modal component :+1:

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi Team, 
## STATUS: CHANGES REQUIRED :recycle: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
almost

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by OmarMWarraich** [comment]
@zakiyousufi kindly let us know if you need our input.

**Comment by zakiyousufi** [comment]
@OmarMWarraich 
Kindly Fix them as I don't have access to my PC I can't resolve them

**Comment by OmarMWarraich** [comment]
> @OmarMWarraich Kindly Fix them as I don't have access to my PC I can't resolve them

ok

**Comment by OmarMWarraich** [comment]
@zakiyousufi @Profsain
The way the modal has been implemented here is not the right way.
I am going to make some changes now. 
Kindly be patient and keep your fingers crossed.
PS: I will delete this message as its only meant for team consumption and I  am assuming that you will be notified in your inboxes once I comment.

**Comment by OmarMWarraich** [comment]
@zakiyousufi @Profsain 
Guys I am working on the modal. 

The problem is that the bootstrap in this condition is always going to create problems for us. Because this will never be a production level feature for our app. 

Having said that, there is a lot in our app that needs to be fixed. 
I mean we could have created a tailwindcss modal once we had tailwindcss however then again tailwind css is never an ideal choice for production.

But these are the things we ought to have discussed in our prep session but as far as I can remember we never had one.

To sum it up, we still have 10 hours before the final deadline. So I would go back to the task , **_fixing the modal and the shopping list table only_**, and hopefully render a solution within the next couple. 

**_Meanwhile, if somebody comes up with another solution, kindly let us know here. The push that I made earlier is just an alternative for Bootstrap cdn so do not mind about it at all._**

**Comment by zakiyousufi** [comment]
Hello @Profsain @OmarMWarraich 
Did our project fix or no because we don't have much time

**Comment by OmarMWarraich** [comment]
> Hello @Profsain @OmarMWarraich Did our project fix or no because we don't have much time

So far we are all working at our ends. Please do not hesitate to feed your input. Issue will be resolved soon. But you are also requested to keep on working at your end as time is of essence.

**Comment by zakiyousufi** [comment]
@OmarMWarraich Hi again 
Please don't push anything to this branch @Profsain is about to fix all errors so please hold on and don't push new commits thank you

**Review by BenMukebo — APPROVED** [review]
# Status: Approved  :champagne:  ✔️:confetti_ball: 

## Hi @TEAM, 

### **Congratulations! 🎉 Project Approved ✅**

Good job so far. Well Done.👍🏻 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

### To Highlight 🎉
- Git flow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

- [OPTIONAL] I would also recommend adding the issue link in the contributing sections: 

```
## 🤝 Contributing

Contributions, issues, and feature requests are welcome!

Feel free to check the [issues page](../../issues/).
```

- [OPTIONAL] Please update your Readme, it will be nice to update this section(`License`) with the type of License and a link to the license file, as this will make your readme more informative and professional.  When I click the link to the `MIT.md` file that you added, I get this 404 `error`. Please, consider fixing the broken link to the MIT.md file 🥇 



Cheers :clinking_glasses:  and Happy coding!!! :dancers:💻  

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by zakiyousufi** [comment]
Hi @BenMukebo @VICTOR-CHUKWU @Kamba56 
Thanks for your kind review and feedback 🙏🏻
